npm/vite-plugin-cypress-esm/src/index.ts

```diff
@@ -0,0 +1,184 @@
+              .map((decl) => decl.replace(/\*\s+as\s+/g, '').trim())
+              .filter((decl) => !!decl),
+            )
+          }
+
+          let destructuredImports = declarations
+          .map((decl) => {
+            // support `import { foo as bar } from 'module'` syntax, converting to `const { foo: bar } ...`
+            decl = decl.replace(/(?<!\*) as /g, ': ')
+
+            return `const ${decl} = ${MODULE_IMPORTER_IDENTIFIER}('${moduleId}', cypress_${moduleIdentifier}_${counter}, ${debug.enabled});`
+          })
+          .join('')
+
+          replacement += destructuredImports
+        })
+
+        return match
+        .replace(importRegex, replacement)
+        .replace(/[\n\r]/g, '')
+        .replace(/const \* as/g, 'const')
+      },
+    )
+  }
+
+  /**
+   * Transform dynamic imports to use a runtime that turns the ES modules into proxies by injecting
+   * a `__cypressDynamicModule` runtime.
+   *
+   * `__cypressDynamicModule` returns a Promise that resolves to the original import which has been
+   * "proxified" to side step the fact ES Modules are immutable and sealed.
+   *
+   * Examples - https://regex101.com/r/Ic1OHA/1
+   *
+   * Given:
+   *   const m = import("./mod_1")
+   *   const m = await import("lodash")
+   *   import("./mod_2").then(mod => mod)
+   *
+   * Returns:
+   *   const m = __cypressDynamicModule(import("./mod_1"))
+   *   const m = await __cypressDynamicModule(import("lodash"))
+   *   __cypressDynamicModule(import("./mod_2")).then(mod => mod)
+   */
+  const mapDynamicImportsToCache = (id: string, code: string) => {
+    const RE = /(import\((.+?)\))/g
+
+    return code.replace(
+      RE,
+      (match, importVars: string, importTarget: string) => {
+        if (importOnIgnoreList(importTarget)) {
+          return match
+        }
+
+        return match.replace(
+          RE,
+          `${MODULE_DYNAMIC_IMPORTER_IDENTIFIER}('${id}', $1, ${debug.enabled})`,
+        )
+      },
+    )
+  }
+
+  return {
+    name: 'cypress:mocks',
+    enforce: 'post',
+    transform (code, id, options) {
+      // Process all files to remap imports
+      // TODO: Restrict to JS files only? Any potential for .mjs etc?
+      let transformedCode = mapImportsToCache(id, code)
+
+      transformedCode = mapDynamicImportsToCache(id, transformedCode)
+
+      return {
+        code: transformedCode,
+      }
+    },
+    async transformIndexHtml (html) {
+      // Process index.html file to remap imports
+      let transformedHtml = mapImportsToCache('index', html)
+
+      transformedHtml = mapDynamicImportsToCache('index', transformedHtml)
+
+      return {
+        html: transformedHtml,
+        tags: [
+          // Pull in Cypress module cache script
+          {
+            tag: 'script',
+            attrs: {
+              type: 'module',
+            },
+            // eslint-disable-next-line no-restricted-syntax
+            children: fs.readFileSync(MODULE_CACHE_FILEPATH, 'utf-8'),
```

Comment:
Is there a reason we ignore this rule here? We say `Synchronous fs calls should not be used in Cypress. Use an async API instead.` so might worth commenting `using readFileSync because x`.